LFS-426_LFS-481: As an administrator i can specify the subject types allowed for a questionnaire in the questionnaire designer

nela-jankechova opened the pull request:
This branch is based off LFS-476, with changes from LFS-426 / LFS-431 & LFS-432 pulled in.

nela-jankechova committed:
LFS-426_LFS-481: Create ArrayInput field

nela-jankechova committed:
LFS-426_LFS-481: rename to 'Fields Input'

nela-jankechova committed:
LFS-426_LFS-481: Get Subject Types

nela-jankechova committed:
LFS-426_LFS-481: Update ListInput

nela-jankechova committed:
LFS-426_LFS-481: Add key to hidden input

sashaandjic commented:
I am not sure how to test this PR. When I click on Subject Types I get "Individual SubjectType view not yet implemented."
![image](https://user-images.githubusercontent.com/23243244/87347761-ba651200-c521-11ea-9651-f2969e756b18.png)

nela-jankechova commented:
> I am not sure how to test this PR. When I click on Subject Types I get "Individual SubjectType view not yet implemented."

@sashaandjic This change is in the questionnaire designer. 

You should be able to go to a questionnaire and see `requiredSubjectTypes` under the `Questionnaire Properties` subheader.  From there, you should be able to click the `edit` button and change the `requiredSubjectTypes` property. The available subject types in the dropdown for you to choose from should correspond to the subject types currently in the database.

I'm currently building this branch so I can see what is displayed for me on the `SubjectTypes` page, but unless this PR broke the subject Type view then it is probably out of scope. 

I hope that helps!

nela-jankechova commented:
Also, here is what i see when I navigate to the `SubjectTypes` tab.

<img width="1061" alt="subjectTypes" src="https://user-images.githubusercontent.com/56768375/87440471-04043a00-c5c0-11ea-8807-3a28a7a43959.PNG">

sashaandjic commented:
I think the functionality within this PR is working as expected.

sashaandjic reviewed (approved):
tested

nela-jankechova committed:
LFS-426_LFS-481: Fix bug

sashaandjic commented:
@nela-jankechova  What was the problem? I want to re-test it.

nela-jankechova commented:
> @nela-jankechova What was the problem? I want to re-test it.

The dropdown was automatically highlighting the first option, rather than the one that was actually selected.

sashaandjic commented:
> > @nela-jankechova What was the problem? I want to re-test it.
> 
> The dropdown was automatically highlighting the first option, rather than the one that was actually selected.

I see. e.g.  in the  Radio Therapy data Tumor region is the only selection. 
![image](https://user-images.githubusercontent.com/23243244/87576545-8e6f9b00-c69f-11ea-90f8-8439169b197a.png)
When I go to edit mode it appears as Patient is the selection, since it is the firt one listed.
![image](https://user-images.githubusercontent.com/23243244/87576485-7a2b9e00-c69f-11ea-8df5-4c73c9489359.png)

sashaandjic commented:
The bug is fixed 
![image](https://user-images.githubusercontent.com/23243244/87583100-73098d80-c6a9-11ea-9b02-b209aa7bb04f.png)

however, after save I need to refresh in order to see my modified selection. Step to reproduce:
- modify existing selection "Tumour region" to "Patient"
![image](https://user-images.githubusercontent.com/23243244/87583269-b06e1b00-c6a9-11ea-91cd-704854170a4a.png)

- click SAVE
- the form is still showing the old selection
![image](https://user-images.githubusercontent.com/23243244/87583332-c7147200-c6a9-11ea-88f3-5c41e50e7b61.png)
- refresh, all looks good
![image](https://user-images.githubusercontent.com/23243244/87583380-dabfd880-c6a9-11ea-8b20-55a725ed22be.png)



-

nela-jankechova commented:
@sashaandjic , this is the same issue as the one you commented for LFS-172. The questionnaire page doesn't update immediately, I thought it was a big enough change to existing code to be its own issue. If it needs fixing before LFS-172 can get merged, then I will do it on that PR.

sashaandjic commented:
Regression after the latest changes. The edit icon is no longer available for Questionnaire Properties. Here is the screenshot showing an older instance built on July 15 and newer one built Aug 28. 
![image](https://user-images.githubusercontent.com/23243244/91625346-98f6a300-e974-11ea-92f0-39538f552be7.png)

OrdiNeu committed:
LFS-426 / LFS-481: As an administrator i can specify the subject types
allowed for a questionnaire in the questionnaire designer
Fix several things that broke in the rebase.
Add a "EditorInput" to consolidate reused code.
Fix bugs in the ListInput handler.

OrdiNeu committed:
LFS-426 / LFS-481: As an administrator i can specify the subject types
allowed for a questionnaire in the questionnaire designer
Fix a warning in React about the useage of keys

OrdiNeu commented:
@sashaandjic It was being fixed at that time. Please re-test.

sashaandjic commented:
I see the pencil button now, however when I create a new questionnaire and click on it to define the type of subject and hit SAVE I get a blank screen
![image](https://user-images.githubusercontent.com/23243244/91891423-00f00680-ec5f-11ea-88b6-29cd4a146e49.png)

sashaandjic commented:
Should minAnswers > 0 be treated as required value? 
If so, I created a task for it https://phenotips.atlassian.net/jira/software/c/projects/LFS/issues/LFS-542

sashaandjic commented:
I can specify number of forms per subject.

sashaandjic reviewed (approved):
tested
I can define one or more subject types for a specific questionnaire.
One small complain: since I am not allowed to save an empty type of subject, it would be nice to have an error message.

sdumitriu reviewed (changes requested):
Unselecting all subject types doesn't work, after saving the previous values are still there.

OrdiNeu committed:
LFS-426 / LFS-481: As an administrator i can specify the subject types
allowed for a questionnaire in the questionnaire designer
Allow list inputs to empty themselves

OrdiNeu commented:
> Unselecting all subject types doesn't work, after saving the previous values are still there.

@sdumitriu I've fixed it in `444af9a` :heavy_check_mark: . It should work now.

sashaandjic commented:
confirming bug reported by @sdumitriu

sashaandjic commented:
tested
